src/Tablebot/Plugins/Quote.hs

```diff
@@ -47,34 +71,143 @@ quoteComm ::
+      qu <- get $ head key
+      case qu of
+        Just q -> renderQuoteMessage q (fromSqlKey $ head key) m
+        Nothing -> sendMessage m "Couldn't find any quotes!"
+      return ()
+  where
+    allQuotes :: [Filter Quote]
+    allQuotes = []
+
+-- | @authorQuote@, which looks for a message of the form @!quote user u@,
+-- selects a random quote from the database attributed to u and responds with that quote.
+-- This is currently mis-documented in the help pages as its not quite working (the author matching is way to strict)
+authorQ :: String -> Message -> DatabaseDiscord ()
+authorQ t m = do
+  num <- count userQuotes
+  if num == 0
+    then sendMessage m "Couldn't find any quotes matching that user!"
+    else do
+      rindex <- liftIO $ randomRIO (0, (num - 1))
+      key <- selectKeysList userQuotes [OffsetBy rindex, LimitTo 1]
+      qu <- get $ head key
+      case qu of
+        Just q -> renderQuoteMessage q (fromSqlKey $ head key) m
+        Nothing -> sendMessage m "Couldn't find any quotes matching that user!"
+      return ()
+  where
+    userQuotes :: [Filter Quote]
+    userQuotes = [QuoteAuthor ==. t]
+
+-- | @addQuote@, which looks for a message of the form
+-- @!quote add "quoted text" - author@, and then stores said quote in the
+-- database, returning the ID used.
+addQ :: String -> String -> Message -> DatabaseDiscord ()
+addQ qu author m = do
+  now <- liftIO $ systemToUTCTime <$> getSystemTime
+  let new = Quote qu author (Just $ fromIntegral $ messageId m) (Just $ fromIntegral $ messageChannel m) now
+  added <- insert $ new
+  let res = pack $ show $ fromSqlKey added
+  renderCustomQuoteMessage ("Quote added as #" `append` res) new (fromSqlKey added) m
+
+-- | @thisQuote@, which takes the replied message or the
+-- previous message and stores said message as a quote in the database,
+-- returning the ID used.
+thisQ :: Message -> DatabaseDiscord ()
+thisQ m = do
+  q <- getReplyMessage m
+  case q of
+    (Just q') -> addMessageQuote q' m
+    Nothing -> do
+      q2 <- getPrecedingMessage m
+      case q2 of
+        (Just q') -> addMessageQuote q' m
+        Nothing -> sendMessage m "Unable to add quote"
+
+-- | @addMessageQuote@, adds a message as a quote to the database, checking that it passes the relevant tests
+addMessageQuote :: Message -> Message -> DatabaseDiscord ()
+addMessageQuote q' m = do
+  num <- count [QuoteMsgId ==. Just (fromIntegral $ messageId q')]
+  if num == 0
+    then
+      if not $ userIsBot (messageAuthor q')
+        then do
+          now <- liftIO $ systemToUTCTime <$> getSystemTime
+          let new = Quote (unpack $ messageText q') (toMentionStr $ messageAuthor q') (Just $ fromIntegral $ messageId q') (Just $ fromIntegral $ messageChannel q') now
+          added <- insert $ new
+          let res = pack $ show $ fromSqlKey added
+          renderCustomQuoteMessage ("Quote added as #" `append` res) new (fromSqlKey added) m
```

Comment:
I think I've seen this come up a few times - might be worth having it as a separate function? Unsure.